services/src/ogc/wms/request.rs

```diff
@@ -0,0 +1,199 @@
+    // sources: - http://docs.geoserver.org/2.8.x/en/user/services/wms/time.html#wms-time
+    //          - http://www.ogcnetwork.net/node/178
+
+    // TODO: support relative time intervals
+
+    let s = <&str as serde::Deserialize>::deserialize(deserializer)?;
+
+    let split: Vec<_> = s.split('/').map(|s| chrono::DateTime::parse_from_rfc3339(s)).collect();
+
+    match *split.as_slice() {
+        [Ok(time)] => TimeInterval::new(time.timestamp(), time.timestamp())
+            .map(Some)
+            .map_err(|_| D::Error::custom("Not a valid time")),
+        [Ok(start), Ok(end)]  => TimeInterval::new(start.timestamp(), end.timestamp())
+            .map(Some)
+            .map_err(|_| D::Error::custom("Not a valid time")),
+        _ => Err(D::Error::custom("Not a valid time"))
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use geoengine_datatypes::primitives::{BoundingBox2D, Coordinate2D};
+
+    #[test]
+    fn deserialize_get_map() {
+        let query = "request=GetMap&service=WMS&version=1.3.0&layer=test&bbox=1,2,3,4&width=2&height=2&crs=foo&styles=ssss&format=image/png&time=2000-01-01T00:00:00.0Z/2000-01-02T00:00:00.0Z";
+        let parsed: WMSRequest = serde_urlencoded::from_str(query).unwrap();
+
+        let request = WMSRequest::GetMap(GetMap {
+            version: "1.3.0".into(),
+            width: 2,
+            layer: "test".into(),
+            crs: "foo".to_string(),
+            styles: "ssss".to_string(),
+            time: Some(TimeInterval::new(946_684_800, 946_771_200).unwrap()),
+            transparent: None,
+            bgcolor: None,
+            sld: None,
+            sld_body: None,
+            elevation: None,
+            bbox: BoundingBox2D::new(Coordinate2D::new(1., 2.),
+                                     Coordinate2D::new(3., 4.)).unwrap(),
+            height: 2,
+            format: GetMapFormat::ImagePng,
+            exceptions: None,
+        });
+
+        assert_eq!(parsed, request);
+    }
+
+    #[test]
+    fn deserialize_get_map_not_time() {
+        let query = "request=GetMap&service=WMS&version=1.3.0&layer=test&bbox=1,2,3,4&width=2&height=2&crs=foo&styles=ssss&format=image/png";
+        let parsed: WMSRequest = serde_urlencoded::from_str(query).unwrap();
+
+        let request = WMSRequest::GetMap(GetMap {
+            version: "1.3.0".into(),
+            width: 2,
+            layer: "test".into(),
+            crs: "foo".to_string(),
+            styles: "ssss".to_string(),
+            time: None,
+            transparent: None,
+            bgcolor: None,
+            sld: None,
+            sld_body: None,
+            elevation: None,
+            bbox: BoundingBox2D::new(Coordinate2D::new(1., 2.),
+                                     Coordinate2D::new(3., 4.)).unwrap(),
+            height: 2,
+            format: GetMapFormat::ImagePng,
+            exceptions: None,
+        });
+
+        assert_eq!(parsed, request);
+    }
```

Comment:
Even if it is tedious, there should be at least a test that has values everywhere.

Reply:
139e2fba1ff1f1d9e93408358e758a05c15f2f12